src/plugin/vk-wrapper/instance/Instance.cpp

```diff
@@ -0,0 +1,304 @@
+
+    _graphicsPipeline.create(device, extent, _renderPass.get());
+
+    auto renderPass = _renderPass.get();
+
+    Framebuffer::CreateInfo framebufferInfo{};
+    framebufferInfo.swapChainExtent = extent;
+    framebufferInfo.renderPass = renderPass;
+    framebufferInfo.swapChainImageViews = _imageView.getImageViews();
+
+    _framebuffer.create(device, framebufferInfo);
+
+    Command::CreateInfo commandInfo{};
+    commandInfo.physicalDevice = _physicalDevice.get();
+    commandInfo.surface = _surface.get();
+    commandInfo.swapChainExtent = extent;
+    commandInfo.renderPass = renderPass;
+    commandInfo.swapChainFramebuffers = _framebuffer.getSwapChainFramebuffers();
+    commandInfo.graphicsPipeline = _graphicsPipeline.get();
+
+    _command.create(device, commandInfo);
+}
+
+void Instance::createSyncObjects()
+{
+    _imageAvailableSemaphores.resize(MAX_FRAMES_IN_FLIGHT);
+    _renderFinishedSemaphores.resize(MAX_FRAMES_IN_FLIGHT);
+    _inFlightFences.resize(MAX_FRAMES_IN_FLIGHT);
+
+    VkSemaphoreCreateInfo semaphoreInfo{};
+    semaphoreInfo.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO;
+
+    VkFenceCreateInfo fenceInfo{};
+    fenceInfo.sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO;
+    fenceInfo.flags = VK_FENCE_CREATE_SIGNALED_BIT;
+
+    auto device = _logicalDevice.get();
+
+    for (uint32_t i = 0; i < MAX_FRAMES_IN_FLIGHT; ++i)
+    {
+        if (vkCreateSemaphore(device, &semaphoreInfo, nullptr, &_imageAvailableSemaphores[i]) != VK_SUCCESS ||
+            vkCreateSemaphore(device, &semaphoreInfo, nullptr, &_renderFinishedSemaphores[i]) != VK_SUCCESS ||
+            vkCreateFence(device, &fenceInfo, nullptr, &_inFlightFences[i]) != VK_SUCCESS)
+            throw std::runtime_error("failed to create semaphores!");
+    }
+}
+
+void Instance::recreateSwapChain(uint32_t width, uint32_t height)
+{
+    auto device = _logicalDevice.get();
+
+    vkDeviceWaitIdle(device);
+
+    cleanupSwapChain(device);
+
+    createSwapChainImages(width, height);
+
+    Framebuffer::CreateInfo framebufferInfo{};
+    framebufferInfo.swapChainExtent = _swapChain.getExtent();
+    framebufferInfo.renderPass = _renderPass.get();
+    framebufferInfo.swapChainImageViews = _imageView.getImageViews();
+
+    _framebuffer.create(device, framebufferInfo);
+}
+
+void Instance::cleanupSwapChain(const VkDevice device)
+{
+    _framebuffer.destroy(device);
+    _imageView.destroy(device);
+    _swapChain.destroy(device);
+}
+
+Result Instance::drawNextImage()
+{
+    auto device = _logicalDevice.get();
+
+    vkWaitForFences(device, 1, &_inFlightFences[_currentFrame], VK_TRUE, UINT64_MAX);
+
+    uint32_t imageIndex;
+    VkResult result = vkAcquireNextImageKHR(device, _swapChain.get(), UINT64_MAX,
+                                            _imageAvailableSemaphores[_currentFrame], VK_NULL_HANDLE, &imageIndex);
+
+    if (result == VK_ERROR_OUT_OF_DATE_KHR)
+        return Result::Failure;
```

Comment:
why are you mixing throwing errors and returnling error codes ?

Reply:
the documentation has been updated to answer the question. To be more specific, this is not an error that stops the application, but one that requires an action in order to continue.

Reply:
can you rename the "Failure" enum option to something like "NeedsBufferSwap" ? (just an example, you may have better naming ideas)